server/src/main/java/io/crate/expression/tablefunctions/UnnestFunction.java

```diff
@@ -68,15 +71,39 @@ public static void register(TableFunctionModule module) {
                 )
                 .withTypeVariableConstraints(typeVariable("E"), typeVariableOfAnyType("N"))
                 .withVariableArity(),
-            UnnestTableFunctionImplementation::new
+            (signature, boundSignature) -> {
```

Comment:
previously, we had kind of similar approach with overriding the function info https://github.com/crate/crate/commit/468e81c77d0c7302a2443e10160845aa50b7cdd3#diff-ac060c995181f48a5efb0b2e51fc576dL63-L70

It seems a bit complicated to do it in the type signature parsing itself :/ and such as we had the approach with overwriting before, i've decided to go this way. Also, it is only the unnest and _values that have this issue.

Reply:
Would it be possible to somehow adapt the `Signature.table` constructor to enforce the building of a correct RowType? 

As it is it's rather difficult to register a table function that is actually working correctly

Reply:
but we are not always registering table functions that have the row type as the return type, for instance, https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/expression/tablefunctions/GenerateSeries.java#L72 we register the function with a return type that corresponds to the type of the input arguments so we can use it as scalar and we use properly constructed row type for `TableFunctionImplementation#returnType` in the case when table function used normally. Therefore, I am not sure that we should enforce `Signature.table`, maybe we can add some shortcut for such cases, e.g. `Signature.tableForArgumetsOnly` or smth, but it will be used only for this specific case. (and _values)

https://github.com/crate/crate/pull/10152/files#diff-3b1c265f133d1b9627ea392b630649c2R65-R70
I think we initially split the unnest signature to emphasise case when there is more than one array of any type
so then we have to return the row type. I think we can merge this and previous signatures and add similar logic that we had before which I mentioned in the previous comment to build all the required function info, but we would still have to rewrite the return type of the bound signature and construct the value for `TableFunctionImplementation#returnType`. If we keep the signatures as they are now then another approach would be to implement some kind of signature binding based on the number of arguments or smth like that, that will be used for this particular case, but it doesn't seem to be that straightforward and I am not sure whether it worth it.

Reply:
Could we then maybe extend the table-function base test class to assert that if the aggregation has a RowType with multiple fields, that the signature result-type must also be a row-type and must match?

Reply:
@mfussenegger pushed the fixup